### ISSUE TITLE
Failing on version 1.3

### ISSUE THREAD

**Issue by sQVe** [Askannz/optimus-manager#253]
First and foremost, thank you for an awesome tool :pray: 

**Describe the bug**

`optimus-manager` is failing to start / execute after updating to version `1.3`. The logs are, sadly, not giving much to go on. The previous versions have worked fine.

**System info**
```
os     Arch Linux
host   XPS 15 7590
wm     i3wm
kernel 5.6.11-arch1-1
```

I'm using regular login with `xinitrc` and `startx`. Running version `1.3-1` from the AUR.

**etc/optimus-manager/optimus-manager.conf:**
```config
[optimus]

# This parameter defines the method used to power switch the Nvidia card. See the documentation
# for a complete description of what each value does. Possible values :
#
# - nouveau : load the nouveau module on the Nvidia card.
# - bbswitch : power off the card using the bbswitch module (requires the bbswitch dependency).
# - acpi_call : try various ACPI method calls to power the card on and off (requires the acpi_call dependency)
# - none : do not use an external module for power management. For some laptop models it's preferable to
#           use this option in combination with pci_power_control (see below).
#           With that option set, you can also use the scripts nvidia-enable.sh and nvidia-disable.sh to
#           execute custom commands for power management.
switching=none

# Enable PCI power management in Intel mode.
# This option is incompatible with acpi_call and bbswitch, so it will be ignored in those cases.
pci_power_control=yes

# Remove the Nvidia card from the PCI bus.
# May prevent crashes caused by power switching.
# Ignored if switching=nouveau or switching=bbswitch.
pci_remove=yes

# Reset the Nvidia card at the PCI level before reloading the nvidia module.
# Ensures the card is in a fresh state before reloading the nvidia module.
# May fix some switching issues. Possible values :
#
# - no : does not perform any reset
# - function_level : perform a light "function-level" reset
# - hot_reset : perform a "hot reset" of the PCI bridge. ATTENTION : this method messes with the hardware
#         directly, please read the online documentation of optimus-manager before using it.
#         Also, it will perform a PCI remove even if pci_remove=no.
#
pci_reset=no

# Automatically log out the current desktop session when switching GPUs.
# This feature is currently supported for the following DE/WM :
# KDE Plasma, GNOME, XFCE, Deepin, i3, Openbox, AwesomeWM, bspwm
# If this option is disabled or you use a different desktop environment,
# GPU switching only becomes effective at the next graphical session login.
auto_logout=yes


[intel]

# Driver to use for the Intel GPU. Possible values : modesetting, intel
# To use the intel driver, you need to install the package "xf86-video-intel".
driver=modesetting

# Acceleration method (corresponds to AccelMethod in the Xorg configuration).
# Only applies to the intel driver.
# Possible values : sna, xna
# Leave blank for the default (no option specified)
accel=

# Enable TearFree option in the Xorg configuration.
# Only applies to the intel driver.
# Possible values : yes, no
# Leave blank for the default (no option specified)
tearfree=

# DRI version. Possible values : 2, 3
DRI=3

# Whether or not to enable modesetting for the nouveau driver.
# Does not affect modesetting for the Intel GPU driver !
# This option only matters if you use nouveau as the switching backend.
modeset=yes

[nvidia]

# Whether or not to enable modesetting. Required for PRIME Synchronization (which prevents tearing).
modeset=yes

# Whether or not to enable the NVreg_UsePageAttributeTable option in the Nvidia driver.
# Recommended, can cause poor CPU performance otherwise.
PAT=yes

# DPI value. This will be set using the Xsetup script passed to your login manager.
# It will run the command
# xrandr --dpi <DPI>
# Leave blank for the default (the above command will not be run).
DPI=96

# If you're running an updated version of xorg-server (let's say to get PRIME Render offload enabled),
# the nvidia driver may not load because of an ABI version mismatch. Setting this flag to "yes"
# will allow the loading of the nvidia driver.
ignore_abi=no

# Comma-separated list of Nvidia-specific options to apply.
# Available options :
# - overclocking : enable CoolBits in the Xorg configuration, which unlocks clocking options
#   in the Nvidia control panel.
# - triple_buffer : enable triple buffering.
options=overclocking
```

**Logs**

```sh
% optimus-manager --status
ERROR: a GPU setup was initiated but Xorg post-start hook did not run.
Log at /var/log/optimus-manager/switch/switch-20200511T163522.log
If your login manager is GDM, make sure to follow those instructions:
https://github.com/Askannz/optimus-manager#important--gnome-and-gdm-users
If your display manager is neither GDM, SDDM nor LightDM, or if you don't use one, read the wiki:
https://github.com/Askannz/optimus-manager/wiki/FAQ,-common-issues,-troubleshooting

Cannot execute command because of previous errors.
```

**/var/log/optimus-manager/switch/switch-20200511T163522.log:**
```sh
[15] INFO: # Xorg pre-start hook
[16] INFO: Requested mode is: intel
[16] INFO: Checking for GDM display servers
[761] INFO: Available modules: ['nouveau', 'nvidia', 'nvidia_drm', 'nvidia_modeset', 'nvidia_uvm']
[761] INFO: Unloading modules ['nvidia_drm', 'nvidia_modeset', 'nvidia_uvm', 'nvidia'] (if loaded)
[763] INFO: Removing Nvidia from PCI bus
[782] WARNING: pci_remove is enabled, pci_power_control option ignored.
[798] INFO: Loaded extra Intel Xorg options (0 lines)
[798] INFO: Loaded extra Intel Xorg options (0 lines)
[798] INFO: Writing to /etc/X11/xorg.conf.d/10-optimus-manager.conf
[798] INFO: Writing state {'type': 'pending_post_xorg_start', 'switch_id': '20200511T163522', 'requested_mode': 'intel'}
[799] INFO: Xorg pre-start hook completed successfully.
```

**/var/log/optimus-manager/daemon/daemon-20200511T163521.log:**
```sh
% cat                                                                                             ~
[34] INFO: # Daemon pre-start hook
[34] INFO: Removing /etc/X11/xorg.conf.d/10-optimus-manager.conf (if present)
[36] INFO: Copying /etc/optimus-manager/optimus-manager.conf to /var/lib/optimus-manager/tmp/config_copy.conf
[38] INFO: Startup mode is: intel
[38] INFO: Writing state {'type': 'pending_pre_xorg_start', 'requested_mode': 'intel', 'current_mode': None}
[38] INFO: Daemon pre-start hook completed successfully.
[38] INFO: Calling Xorg pre-start hook.
[1] INFO: # Commands daemon
[1] INFO: Opening UNIX socket
[1] INFO: Awaiting commands
```

**Comment by qumaciel** [comment]
I'm getting the same issue in manjaro.

**Comment by Askannz** [comment]
Actually the logs do tell us something. They're missing the Xorg post-startup part, which means `prime-offload` did not execute. Have you put it in your `.xinitrc` ?

**Comment by sQVe** [comment]
> Actually the logs do tell us something. They're missing the Xorg post-startup part, which means `prime-offload` did not execute. Have you put it in your `.xinitrc` ?

No. I have not changed my setup at all since updating. Just enabling the service worked fine for me previously.

So, I need to enable / start prime off-loading in my `xinitrc`? I'll take a look later today.

**Comment by sQVe** [comment]
So, I managed to get everything working perfectly :+1:

First we need to enable running `prime-switch` as `sudo` without the password prompt by adding via `visudo`:

```
sqve ALL=(ALL) NOPASSWD: /usr/bin/prime-switch
```

The following commits adds the rest:
- https://github.com/sQVe/dotfiles/commit/40ac0fad05ff35fa094e10bb38d338844076730d
- https://github.com/sQVe/scripts/commit/e28aef27862477672f268578fecdd575e529c3d7

**Comment by soumendu041** [comment]
I'm also having the same issue. I am using lightdm and xfce on Arch. My logfiles look similar to sQVe's. Also, `systemctl status optimus-manager.service` says:
```
● optimus-manager.service - Optimus Manager Commands Daemon
     Loaded: loaded (/usr/lib/systemd/system/optimus-manager.service; enabled; vendor preset: disabled)
     Active: active (running) since Tue 2020-05-12 13:53:07 IST; 3min 11s ago
    Process: 1476 ExecStartPre=/usr/bin/python3 -u -m optimus_manager.hooks.pre_daemon_start (code=exited, status=0/SUCCE>
    Process: 1535 ExecStartPre=/usr/bin/python3 -u -m optimus_manager.hooks.pre_xorg_start (code=exited, status=0/SUCCESS)
   Main PID: 1929 (python3)
      Tasks: 1 (limit: 23471)
     Memory: 35.1M
     CGroup: /system.slice/optimus-manager.service
             └─1929 /usr/bin/python3 -u -m optimus_manager.daemon

May 12 13:55:33 archerzeta python3[1929]:     "type": "switch"
May 12 13:55:33 archerzeta python3[1929]: }
May 12 13:55:33 archerzeta python3[1929]: [145749] INFO: Writing requested GPU mode nvidia
May 12 13:55:33 archerzeta python3[1929]: [145749] ERROR: Invalid command  "{
May 12 13:55:33 archerzeta python3[1929]:     "args": {
May 12 13:55:33 archerzeta python3[1929]:         "mode": "nvidia"
May 12 13:55:33 archerzeta python3[1929]:     },
```

**Comment by sQVe** [comment]
@soumendu041 You should probably try `journalctl -u optimus-manager.service -b 0` and provide the output here.

@Askannz I suggest closing this issue as my problem is solved. The documentation should be improved in these edge cases when not running a DE.

**Comment by soumendu041** [comment]
@sQVe Thanks. Here's the output:
```
May 12 20:07:23 archerzeta systemd[1]: Starting Optimus Manager Commands Daemon...
May 12 20:07:23 archerzeta python3[1470]: [45] INFO: # Daemon pre-start hook
May 12 20:07:23 archerzeta python3[1470]: [45] INFO: Removing /etc/X11/xorg.conf.d/10-optimus-manager.conf (if present)
May 12 20:07:23 archerzeta python3[1470]: [46] INFO: Copying /etc/optimus-manager/optimus-manager.conf to /var/lib/optimu>
May 12 20:07:23 archerzeta python3[1470]: [50] INFO: Startup mode is: intel
May 12 20:07:23 archerzeta python3[1470]: [50] INFO: Writing state {'type': 'pending_pre_xorg_start', 'requested_mode': '>
May 12 20:07:23 archerzeta python3[1470]: [50] INFO: Daemon pre-start hook completed successfully.
May 12 20:07:23 archerzeta python3[1470]: [50] INFO: Calling Xorg pre-start hook.
May 12 20:07:23 archerzeta python3[1527]: [12] INFO: # Xorg pre-start hook
May 12 20:07:23 archerzeta python3[1527]: [12] INFO: Requested mode is: intel
May 12 20:07:23 archerzeta python3[1527]: [12] INFO: Checking for GDM display servers
May 12 20:07:24 archerzeta python3[1527]: [868] INFO: Available modules: ['nouveau', 'bbswitch', 'acpi_call', 'nvidia', '>
May 12 20:07:24 archerzeta python3[1527]: [868] INFO: Unloading modules ['nvidia_drm', 'nvidia_modeset', 'nvidia_uvm', 'n>
May 12 20:07:24 archerzeta python3[1527]: [874] INFO: Loading module bbswitch
May 12 20:07:24 archerzeta python3[1527]: [882] INFO: Setting GPU power to OFF via bbswitch
May 12 20:07:24 archerzeta python3[1527]: [1037] INFO: Loaded extra Intel Xorg options (0 lines)
May 12 20:07:24 archerzeta python3[1527]: [1038] INFO: Loaded extra Intel Xorg options (0 lines)
```
Seems like the problem is with `lightdm`. After switching to `gdm-prime` from AUR optimus-manager works just fine.

**Comment by Askannz** [comment]
@sQVe thanks for the detailed report. There's clearly some friction between optimus-manager and environment without display managers, it's something I need to investigate.

@soumendu041 Do you get the same error message as sQVe when running `optimus-manager --status` ? If yes, could you join the attached log ? Thank you.

**Comment by soumendu041** [comment]
@Askannz Yes, here you go!
```
 $ optimus-manager --status                                      
ERROR: a GPU setup was initiated but Xorg post-start hook did not run.
Log at /var/log/optimus-manager/switch/switch-20200513T213222.log
If your login manager is GDM, make sure to follow those instructions:
https://github.com/Askannz/optimus-manager#important--gnome-and-gdm-users
If your display manager is neither GDM, SDDM nor LightDM, or if you don't use one, read the wiki:
https://github.com/Askannz/optimus-manager/wiki/FAQ,-common-issues,-troubleshooting

Cannot execute command because of previous errors.
```
```
$  cat /var/log/optimus-manager/switch/switch-20200513T213222.log 
[13] INFO: # Xorg pre-start hook
[13] INFO: Requested mode is: intel
[14] INFO: Checking for GDM display servers
[870] INFO: Available modules: ['nouveau', 'bbswitch', 'acpi_call', 'nvidia', 'nvidia_drm', 'nvidia_modeset', 'nvidia_uvm']
[870] INFO: Unloading modules ['nvidia_drm', 'nvidia_modeset', 'nvidia_uvm', 'nvidia'] (if loaded)
[876] INFO: Loading module bbswitch
[884] INFO: Setting GPU power to OFF via bbswitch
[1041] INFO: Loaded extra Intel Xorg options (0 lines)
[1041] INFO: Loaded extra Intel Xorg options (0 lines)
[1041] INFO: Writing to /etc/X11/xorg.conf.d/10-optimus-manager.conf
[1041] INFO: Writing state {'type': 'pending_post_xorg_start', 'switch_id': '20200513T213222', 'requested_mode': 'intel'}
[1042] INFO: Xorg pre-start hook completed successfully.
```

**Comment by jamesj2** [comment]
I too ran into this same issue after some updates.  It seems apparmor was causing it to fail.  After stopping the service optimus-manager started working again.

**Comment by 2mc** [comment]
@jamesj2 On archlinux, I have no apparmor enabled, not even installed. I start X manually and, yes, `prime-offload` is run just after the X server is started. Everything work perfectly until this update.

But after a fresh boot even before starting Xorg and while having `optimus-manager.service` running, ` optimus-manager --status` tells me that "[...] Xorg post-start hook did not run. [...]" exactly the same way as posted by @soumendu041 3 days before. Already the kind of message indicates that something must be buggy. 

Could you please elaborate, how you got the great `optimus-manager` working again?

**Comment by larsveelaert** [comment]
@2mc I just had this issue and fixed it like this;
After closing the X server, your machine needs to run `sudo prime-switch`.

I did this by adding it at the end of the .xinit file (note: do not use exec, so the execution is halted)
```
...
prime-offload &
openbox
sudo prime-switch
```

Doing this approach you'd have to add the following line to your sudo-config;
```
%wheel ALL=(ALL) NOPASSWD: /usr/bin/prime-switch
```
This way you can run sudo commands without the password prompt for this specific command

Still in progress of smoothing it out, but it works for me now.

**Comment by mechaprabal** [comment]
Not able to switch from `intel` mode to any other mode.
**Sys info:**
```
XPS 7590
Arch Linux + KDE

Clean install of optimus-manager v1.3
```
**Logs:**
```
> optimus-manager --status
ERROR: the latest GPU setup attempt failed at Xorg pre-start hook.
Log at /var/log/optimus-manager/switch/switch-20200517T080919.log

Cannot execute command because of previous errors.
```
```
> /var/log/optimus-manager/switch/switch-20200517T080919.log
[5] INFO: # Xorg pre-start hook
[5] INFO: Requested mode is: nvidia
[5] INFO: Checking for GDM display servers
[676] INFO: Available modules: ['nouveau', 'nvidia', 'nvidia_drm', 'nvidia_modeset', 'nvidia_uvm']
[676] INFO: Unloading modules ['nouveau'] (if loaded)
[3610] INFO: Nvidia card not visible in PCI bus, rescanning
[3610] INFO: Rescanning PCI bus
[3682] ERROR: Nvidia card not showing up in PCI bus after rescan. Continuing anyways.
[3682] INFO: Setting Nvidia PCI power state to on
[3802] ERROR: Cannot set PCI power management state. Continuing anyways. Error is: Nvidia not in PCI bus
[3827] INFO: Loading module nvidia
[4666] ERROR: Xorg pre-start setup error
Traceback (most recent call last):
  File "/usr/lib/python3.8/site-packages/optimus_manager/bash.py", line 11, in exec_bash
    out = subprocess.check_output(
  File "/usr/lib/python3.8/subprocess.py", line 411, in check_output
    return run(*popenargs, stdout=PIPE, timeout=timeout, check=True,
  File "/usr/lib/python3.8/subprocess.py", line 512, in run
    raise CalledProcessError(retcode, process.args,
subprocess.CalledProcessError: Command '['bash', '-c', 'modprobe nvidia NVreg_UsePageAttributeTable=1']' returned non-zero exit status 1.

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/usr/lib/python3.8/site-packages/optimus_manager/kernel.py", line 179, in _load_module
    exec_bash("modprobe %s %s" % (module, options))
  File "/usr/lib/python3.8/site-packages/optimus_manager/bash.py", line 18, in exec_bash
    raise BashError(
optimus_manager.bash.BashError: Failed to execute 'modprobe nvidia NVreg_UsePageAttributeTable=1' :
modprobe: ERROR: could not insert 'nvidia': No such device


During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/usr/lib/python3.8/site-packages/optimus_manager/hooks/pre_xorg_start.py", line 33, in main
    setup_kernel_state(config, prev_state, requested_mode)
  File "/usr/lib/python3.8/site-packages/optimus_manager/kernel.py", line 21, in setup_kernel_state
    _nvidia_up(config)
  File "/usr/lib/python3.8/site-packages/optimus_manager/kernel.py", line 56, in _nvidia_up
    _load_nvidia_modules(config, available_modules)
  File "/usr/lib/python3.8/site-packages/optimus_manager/kernel.py", line 109, in _load_nvidia_modules
    _load_module(available_modules, "nvidia", options="NVreg_UsePageAttributeTable=%d" % pat_value)
  File "/usr/lib/python3.8/site-packages/optimus_manager/kernel.py", line 181, in _load_module
    raise KernelSetupError("error running modprobe for %s : %s" % (module, str(e)))
optimus_manager.kernel.KernelSetupError: error running modprobe for nvidia : Failed to execute 'modprobe nvidia NVreg_UsePageAttributeTable=1' :
modprobe: ERROR: could not insert 'nvidia': No such device

[4669] INFO: Removing /etc/X11/xorg.conf.d/10-optimus-manager.conf (if present)
[4669] INFO: Writing state {'type': 'pre_xorg_start_failed', 'switch_id': '20200517T080919', 'requested_mode': 'nvidia'}

```

Please help making optimus-manager start working again on this system.

**Comment by Askannz** [comment]
@mechaprabal Looks like a typical case of the Nvidia GPU not coming back online. To confirm it, try disabling all power management options. Are you sure switching from intel to nvidia was working with version 1.2.2, with the same power management options enabled ?
In any case, it's a different problem, so please open a new issue.

@2mc how do you run `prime-offload` ?

@qumaciel what is your display manager ? if you don't use one, make sure `prime-offload` and `prime-switch` and run as explained by @sQVe and @larsveelaert.

@soumendu041 I can't reproduce the issue, lightdm works fine with optimus-manager here. Do you have custom configuration in `/etc/lightdm/lightdm.conf` or `/etc/lightdm/lightdm.conf.d/` that would override the property `display-setup-script` set by optimus-manager ?

**Comment by soumendu041** [comment]
@Askannz Spot on! I had a custom `display-setup-script` lingering around from the dark times when optimus-manager did not exist (it did not cause a problem under optimus-manager 1.2). I can confirm that optimus-manager 1.3 works as expected if I remove the custom script. It would be nice if you could add this as a note in the README for the benefit of new users.

**Comment by mechaprabal** [comment]
@Askannz Version 1.22 was working exceptionally well with all the modes.
I do not have any power management scripts in use.
Please direct me with the commands you would like me to run.
Thanks!

**Comment by 2mc** [comment]
@larsveelaert Thank you so much for your elaboration! I did not use `exec` but I sourced a script from `.xinitrc` which checked if `optimus-manager` was active and then called `prime-offlload` as so:

    if systemctl -q is-active "optimus-manager" ; then
      [ -x /usr/bin/prime-offload ] && . /usr/bin/prime-offload
    fi

This 'context-aware' way used to work perfectly before. NOW, however, thanks to your hint, I put `prime-offload` just in front of the X-program like so: `prime-offload && startplasma-x11` And, oh wonder, now it works again!

@Askannz Thank you for reading this bug report thread to carefully! I think, I answered your question right above. However, I liked the former solution much more. Is there a reason for the change of behavior, or is there a way to get the old, more convenient one, back again?

**Comment by mechaprabal** [comment]
@Askannz Now I can switch to all the modes after editing `pci_remove=no` in the config. Further I implemented the dynamic power management as advised in this [link](http://download.nvidia.com/XFree86/Linux-x86_64/435.17/README/dynamicpowermanagement.html).  
**Issue:**  

In hybrid mode, the `nvidia` card is serving the `Xorg` which keeps the card switched On.  The dynamic power management is not able to switch off the card if the Xorg keeps getting served by nvidia.
I tried to make `Intel iGPU` run the Xorg by providing configuration in `/etc/X11/xorg.conf.d/10-nvidia-drm-outputclass.conf` but the NVIDIA GPU is still running the Xorg (verified by `nvidia-smi`)

Can someone please help me to run Xorg on intel instead of Nvidia in the `hybrid` mode?

**Comment by ac-buftea** [comment]
@Askannz I can confirm the same as @soumendu041 . I use LightDM with xfce and have a setup script (`display-setup-script=/usr/bin/numlockx on`) to enable numlock at startup. After the update, changing the gpu works only if I comment this line and disable the script. I've tested it multiple times.

PC:
```
OS: Arch Linux x86_64
Host: Asus G56JK 1.0
Kernel: 5.6.13-arch1-1
```

Error:
```
ERROR: a GPU setup was initiated but Xorg post-start hook did not run.
Log at /var/log/optimus-manager/switch/switch-20200521T180712.log
If your login manager is GDM, make sure to follow those instructions:
https://github.com/Askannz/optimus-manager#important--gnome-and-gdm-users
If your display manager is neither GDM, SDDM nor LightDM, or if you don't use one, read the wiki:
https://github.com/Askannz/optimus-manager/wiki/FAQ,-common-issues,-troubleshooting

Cannot execute command because of previous errors.
```

**Comment by EPadronU** [comment]
I'm experimenting the same.

### optimus-manager --print-mode
```
ERROR: a GPU setup was initiated but Xorg post-start hook did not run.
Log at /var/log/optimus-manager/switch/switch-20200606T072048.log
If your login manager is GDM, make sure to follow those instructions:
https://github.com/Askannz/optimus-manager#important--gnome-and-gdm-users
If your display manager is neither GDM, SDDM nor LightDM, or if you don't use one, read the wiki:
https://github.com/Askannz/optimus-manager/wiki/FAQ,-common-issues,-troubleshooting

Cannot execute command because of previous errors.
```

### cat /var/log/optimus-manager/switch/switch-20200606T072048.log                                                                                            

```
[11] INFO: # Xorg pre-start hook
[11] INFO: Requested mode is: intel
[11] INFO: Checking for GDM display servers
[271] INFO: Available modules: ['nouveau', 'bbswitch', 'acpi_call', 'nvidia', 'nvidia_drm', 'nvidia_modeset', 'nvidia_uvm']
[271] INFO: Unloading modules ['nvidia_drm', 'nvidia_modeset', 'nvidia_uvm', 'nvidia'] (if loaded)
[292] INFO: Loaded extra Intel Xorg options (0 lines)
[293] INFO: Loaded extra Intel Xorg options (0 lines)
[293] INFO: Writing to /etc/X11/xorg.conf.d/10-optimus-manager.conf
[293] INFO: Writing state {'type': 'pending_post_xorg_start', 'switch_id': '20200606T072048', 'requested_mode': 'intel'}
[294] INFO: Xorg pre-start hook completed successfully.
```

### uname

Linux 5.6.15-1-MANJARO #1 SMP PREEMPT Wed May 27 20:38:56 UTC 2020 x86_64 GNU/Linux

### Desktop environment / Display Manager

KDE Plasma / SDDM

### Config

```
[optimus]
switching=bbswitch
pci_power_control=no
pci_remove=no
pci_reset=function_level
auto_logout=no
startup_mode=nvidia # Same as for intel
startup_auto_battery_mode=intel
startup_auto_extpower_mode=nvidia
[intel]
driver=intel
accel=sna
tearfree=yes
DRI=3
modeset=yes
[nvidia]
modeset=yes
PAT=yes
DPI=96
ignore_abi=no
allow_external_gpus=no
options=overclocking
```

### journalctl -u sddm.service
```
jun 06 07:20:48 aincrad systemd[1]: Started Simple Desktop Display Manager.
jun 06 07:20:49 aincrad sddm[1005]: Initializing...
jun 06 07:20:49 aincrad sddm[1005]: Starting...
jun 06 07:20:49 aincrad sddm[1005]: Logind interface found
jun 06 07:20:49 aincrad sddm[1005]: Adding new display on vt 1 ...
jun 06 07:20:49 aincrad sddm[1005]: Loading theme configuration from ""
jun 06 07:20:49 aincrad sddm[1005]: Display server starting...
jun 06 07:20:49 aincrad sddm[1005]: Running: /usr/bin/X -dpi 0 -auth /var/run/sddm/{17295d04-4b0c-442b-8ce7-e4e39abbced9} -background none -noreset -displayfd 17 -seat seat0 vt1
jun 06 07:20:51 aincrad sddm[1005]: Setting default cursor
jun 06 07:20:51 aincrad sddm[1005]: Running display setup script  "/usr/share/sddm/scripts/Xsetup"
jun 06 07:20:51 aincrad sddm[1005]: Display server started.
jun 06 07:20:51 aincrad sddm[1005]: Socket server starting...
jun 06 07:20:51 aincrad sddm[1005]: Socket server started.
jun 06 07:20:51 aincrad sddm[1005]: Loading theme configuration from "/usr/share/sddm/themes/breath/theme.conf"
jun 06 07:20:51 aincrad sddm[1005]: Greeter starting...
jun 06 07:20:51 aincrad sddm[1005]: Adding cookie to "/var/run/sddm/{17295d04-4b0c-442b-8ce7-e4e39abbced9}"
jun 06 07:20:51 aincrad sddm-helper[1051]: [PAM] Starting...
jun 06 07:20:51 aincrad sddm-helper[1051]: [PAM] Authenticating...
jun 06 07:20:51 aincrad sddm-helper[1051]: [PAM] returning.
jun 06 07:20:51 aincrad sddm-helper[1051]: pam_unix(sddm-greeter:session): session opened for user sddm by (uid=0)
jun 06 07:20:51 aincrad sddm[1005]: Greeter session started successfully
jun 06 07:20:51 aincrad sddm[1005]: Message received from greeter: Connect
jun 06 07:20:56 aincrad sddm[1005]: Message received from greeter: Login
jun 06 07:20:56 aincrad sddm[1005]: Reading from "/usr/share/xsessions/plasma.desktop"
jun 06 07:20:56 aincrad sddm[1005]: Reading from "/usr/share/xsessions/plasma.desktop"
jun 06 07:20:56 aincrad sddm[1005]: Session "/usr/share/xsessions/plasma.desktop" selected, command: "/usr/bin/startplasma-x11"
jun 06 07:20:56 aincrad sddm-helper[1321]: [PAM] Starting...
jun 06 07:20:56 aincrad sddm-helper[1321]: [PAM] Authenticating...
jun 06 07:20:56 aincrad sddm-helper[1321]: [PAM] Preparing to converse...
jun 06 07:20:56 aincrad sddm-helper[1321]: [PAM] Conversation with 1 messages
jun 06 07:20:56 aincrad sddm-helper[1321]: pam_kwallet5(sddm:auth): (null): pam_sm_authenticate
jun 06 07:20:56 aincrad sddm-helper[1321]: [PAM] returning.
jun 06 07:20:56 aincrad sddm[1005]: Authenticated successfully
jun 06 07:20:56 aincrad sddm-helper[1321]: pam_kwallet5(sddm:setcred): pam_kwallet5: pam_sm_setcred
jun 06 07:20:56 aincrad sddm-helper[1321]: pam_unix(sddm:session): session opened for user nodxine by (uid=0)
jun 06 07:20:56 aincrad sddm-helper[1321]: pam_kwallet5(sddm:session): pam_kwallet5: pam_sm_open_session
jun 06 07:20:56 aincrad sddm[1005]: Auth: sddm-helper exited successfully
jun 06 07:20:56 aincrad sddm[1005]: Greeter stopped.
jun 06 07:20:56 aincrad sddm-helper[1321]: Starting: "/usr/share/sddm/scripts/Xsession \"/usr/bin/startplasma-x11\""
jun 06 07:20:56 aincrad sddm[1005]: Session started
```

### journalctl -u optimus-manager.service -b 0 

```
jun 06 07:20:47 aincrad systemd[1]: Starting Optimus Manager Commands Daemon...
jun 06 07:20:48 aincrad python3[894]: [45] INFO: # Daemon pre-start hook
jun 06 07:20:48 aincrad python3[894]: [46] INFO: Removing /etc/X11/xorg.conf.d/10-optimus-manager.conf (if present)
jun 06 07:20:48 aincrad python3[894]: [48] INFO: Startup mode is: intel
jun 06 07:20:48 aincrad python3[894]: [48] INFO: Writing state {'type': 'pending_pre_xorg_start', 'requested_mode': 'intel', 'current_mode': None}
jun 06 07:20:48 aincrad python3[894]: [49] INFO: Daemon pre-start hook completed successfully.
jun 06 07:20:48 aincrad python3[894]: [49] INFO: Calling Xorg pre-start hook.
jun 06 07:20:48 aincrad python3[937]: [11] INFO: # Xorg pre-start hook
jun 06 07:20:48 aincrad python3[937]: [11] INFO: Requested mode is: intel
jun 06 07:20:48 aincrad python3[937]: [11] INFO: Checking for GDM display servers
jun 06 07:20:48 aincrad python3[937]: [271] INFO: Available modules: ['nouveau', 'bbswitch', 'acpi_call', 'nvidia', 'nvidia_drm', 'nvidia_modeset', 'nvidia_uvm']
jun 06 07:20:48 aincrad python3[937]: [271] INFO: Unloading modules ['nvidia_drm', 'nvidia_modeset', 'nvidia_uvm', 'nvidia'] (if loaded)
jun 06 07:20:48 aincrad python3[937]: [292] INFO: Loaded extra Intel Xorg options (0 lines)
jun 06 07:20:48 aincrad python3[937]: [293] INFO: Loaded extra Intel Xorg options (0 lines)
jun 06 07:20:48 aincrad python3[937]: [293] INFO: Writing to /etc/X11/xorg.conf.d/10-optimus-manager.conf
jun 06 07:20:48 aincrad python3[937]: [293] INFO: Writing state {'type': 'pending_post_xorg_start', 'switch_id': '20200606T072048', 'requested_mode': 'intel'}
jun 06 07:20:48 aincrad python3[937]: [294] INFO: Xorg pre-start hook completed successfully.
jun 06 07:20:48 aincrad systemd[1]: Started Optimus Manager Commands Daemon.
jun 06 07:20:48 aincrad python3[987]: [1] INFO: # Commands daemon
jun 06 07:20:48 aincrad python3[987]: [1] INFO: Opening UNIX socket
jun 06 07:20:48 aincrad python3[987]: [1] INFO: Awaiting commands
```

### systemctl status optimus-manager.service
```
● optimus-manager.service - Optimus Manager Commands Daemon
     Loaded: loaded (/usr/lib/systemd/system/optimus-manager.service; enabled; vendor preset: disabled)
     Active: active (running) since Sat 2020-06-06 07:20:48 -05; 7min ago
    Process: 894 ExecStartPre=/usr/bin/python3 -u -m optimus_manager.hooks.pre_daemon_start (code=exited, status=0/SUCCESS)
    Process: 937 ExecStartPre=/usr/bin/python3 -u -m optimus_manager.hooks.pre_xorg_start (code=exited, status=0/SUCCESS)
   Main PID: 987 (python3)
      Tasks: 1 (limit: 19049)
     Memory: 32.3M
     CGroup: /system.slice/optimus-manager.service
             └─987 /usr/bin/python3 -u -m optimus_manager.daemon

jun 06 07:20:48 aincrad python3[937]: [271] INFO: Unloading modules ['nvidia_drm', 'nvidia_modeset', 'nvidia_uvm', 'nvidia'] (if loaded)
jun 06 07:20:48 aincrad python3[937]: [292] INFO: Loaded extra Intel Xorg options (0 lines)
jun 06 07:20:48 aincrad python3[937]: [293] INFO: Loaded extra Intel Xorg options (0 lines)
jun 06 07:20:48 aincrad python3[937]: [293] INFO: Writing to /etc/X11/xorg.conf.d/10-optimus-manager.conf
jun 06 07:20:48 aincrad python3[937]: [293] INFO: Writing state {'type': 'pending_post_xorg_start', 'switch_id': '20200606T072048', 'requested_mode': 'intel'}
jun 06 07:20:48 aincrad python3[937]: [294] INFO: Xorg pre-start hook completed successfully.
jun 06 07:20:48 aincrad systemd[1]: Started Optimus Manager Commands Daemon.
jun 06 07:20:48 aincrad python3[987]: [1] INFO: # Commands daemon
jun 06 07:20:48 aincrad python3[987]: [1] INFO: Opening UNIX socket
jun 06 07:20:48 aincrad python3[987]: [1] INFO: Awaiting commands
```

**Comment by Askannz** [comment]
@ac-buftea makes sense, optimus-manager needs that script to run and will not work if you replace it with your own.

The only way to make both work would be to override the `display-setup-script` property with a third script that executes both.

@mechaprabal are you sure that the Xorg server is still running on the Nvidia card in hybrid mode ? What does `optimus-manager --status` say ?

**Comment by 0x192** [comment]
@Askannz 

Thanks a lot for this wonderful tool ! :heart: 

I had the same issue than @mechaprabal. Switching to `optimus-manager-git` fixed it. :+1:

**Comment by smac89** [comment]
I was getting this error with SDDM window manager and the reason was because I had initially followed this [guide](https://wiki.archlinux.org/index.php/NVIDIA_Optimus#SDDM) on Arch wiki to setup SDDM with Nvidia.
Then I guess at some point, I also generated the config file for SDDM using `sddm --example-config | sudo tee /etc/sddm.conf`.

The problem is that `optimus-manager` comes with it's own setup (`prime-offload`) and tear down (`prime-switch`) scripts, and these can be found in `/etc/sddm.conf.d/20-optimus-manager.conf`.

So all I had to do was to open `sddm-config-editor` and remove the scripts that were specified to be executed when the display server starts or stops, then after saving and restarting, I never get this error again.
